### ISSUE TITLE
Connection pooling and connection lifetime

### ISSUE THREAD

**Issue by RoadTrain** [DarkWanderer/ClickHouse.Client#105]
Hello! Thanks for the great library.

I have a several somewhat general questions regarding `ClickHouseConnection` management.

1. What is the preferred lifetime for a connection object? How long can a connection be reused?
2. Is there some sort of [connection pooling](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/connection-pooling) like in SQL Server ADO.NET? 
3. How to check if the connection is still alive?
4. When using a DI framework, what is the best way to create/manage connection instances?

Basically, I'd like to find a way to achieve both efficient resource use and fault tolerance.

**Comment by DarkWanderer** [comment]
Hi @RoadTrain 

Glad it can be useful. On your questions:

1. A `ClickHouseConnection` represents a "session" with the server. It performs feature discovery by querying server version (so there is a minor overhead), but generally it is safe to create and destroy such objects multiple times. A connection object does not 1:1 map to an open TCP connection - they can be multiplexed (see next question)

2. There is implicit connection pooling by default. Underlying connections are backed by `System.Net.HttpClient` class - which implements a per-endpoint connection pool ([more info](https://www.stevejgordon.co.uk/httpclient-connection-pooling-in-dotnet-core)). This means that by default, if you have many open connections to single ClickHouse server, they will be multiplexed through a handful of HTTP keepalive connections. You can tweak this behavior by passing a `HttpClient` with custom `HttpClientHandler` settings

3. There isn't a way to check it other than running a query. But if you did a successful query before, next one will succeed unless server died/password changed/your network died etc. - i.e. you only need to care about handling the actual query error and don't need to try and manage "dead"/"alive" connections

4. Best way would be to create one connection object per large "transaction" spanning multiple queries, then dispose it. If you use ADO.NET DI, there is a constructor which takes `IHttpClientFactory` (make sure to read the docstring on it)

**Comment by RoadTrain** [comment]
Thanks, that's a very detailed explanation! 🎉
Maybe it's worth putting it in Wiki.

**Comment by DarkWanderer** [comment]
Added an article about lifetime and pooling
https://github.com/DarkWanderer/ClickHouse.Client/wiki/Connection-lifetime-&-pooling
